backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/highlighter/impl/LuceneUnifiedSearchHighlighter.java

```diff
@@ -0,0 +1,186 @@
+		);
+	}
+
+	@Override
+	public LuceneAbstractSearchHighlighter withFallbackDefaults() {
+		return withFallback( DEFAULTS );
+	}
+
+	@Override
+	public <A> Values<A> createValues(String parentDocumentPath, String nestedDocumentPath,
+			String absoluteFieldPath, Analyzer analyzer, ProjectionExtractContext context,
+			ProjectionAccumulator<String, ?, A, List<String>> accumulator) {
+		return new UnifiedHighlighterValues<>(
+				parentDocumentPath, nestedDocumentPath, absoluteFieldPath, analyzer, context, accumulator );
+	}
+
+	@Override
+	public SearchHighlighterType type() {
+		return SearchHighlighterType.UNIFIED;
+	}
+
+	@Override
+	public void checkApplicability(LuceneSearchIndexValueFieldTypeContext<?> typeContext) {
+		if ( typeContext.hasTermVectorsConfigured() && this.maxAnalyzedOffset != null && this.maxAnalyzedOffset > 0 ) {
+			throw log.unifiedHighlighterMaxAnalyzedOffsetNotSupported();
+		}
+	}
+
+	private final class UnifiedHighlighterValues<A> extends HighlighterValues<A> {
+
+		private final String[] fieldsIn;
+		private final int[] maxPassagesIn;
+		private final Query query;
+		private final UnifiedHighlighter highlighter;
+
+		UnifiedHighlighterValues(String parentDocumentPath, String nestedDocumentPath, String field, Analyzer analyzer,
+				ProjectionExtractContext context, ProjectionAccumulator<String, ?, A, List<String>> accumulator) {
+			super( parentDocumentPath, nestedDocumentPath, context.collectorExecutionContext(), accumulator );
+			this.fieldsIn = new String[] { field };
+			this.maxPassagesIn = new int[] { LuceneUnifiedSearchHighlighter.this.numberOfFragments };
+			this.query = context.collectorExecutionContext().originalQuery();
+			PassageFormatterWithEncoder formatter = new PassageFormatterWithEncoder(
+					LuceneUnifiedSearchHighlighter.this.preTags.get( 0 ),
+					LuceneUnifiedSearchHighlighter.this.postTags.get( 0 ),
+					LuceneUnifiedSearchHighlighter.this.encoder
+			);
+
+			this.highlighter = new UnifiedHighlighter(
+					new IndexSearcher( context.collectorExecutionContext().getIndexSearcher().getIndexReader() ),
+					LimitTokenOffsetAnalyzer.analyzer( analyzer, LuceneUnifiedSearchHighlighter.this.maxAnalyzedOffset )
+			);
+			highlighter.setFormatter( formatter );
+			highlighter.setBreakIterator( this::breakIterator );
+			// to correctly support no match size we need to override UnifiedHighlighter#getFieldHighlighter() so that
+			// we can return our own custom field highlighter that would cut the text where we need rather than returning
+			// entire field value...
+			if ( LuceneUnifiedSearchHighlighter.this.noMatchSize > 0 && LuceneUnifiedSearchHighlighter.this.noMatchSize != Integer.MAX_VALUE ) {
+				// we don't want to throw an exception here as users might still want to get the no-match fragment even if
+				// they cannot limit the length of it...
+				log.unifiedHighlighterNoMatchSizeWarning( LuceneUnifiedSearchHighlighter.this.noMatchSize );
+			}
```

Comment:
I thought that it might be better to log it when the highlighter is getting used rather than created so that when someone configured one but didn't apply it to the query or to a field (named highlighter), then it won't log ... but now thinking about it ... it would log multiple times - once per field ... so probably it wasn't a good idea after all 😄 will move it.